draft-ietf-tls-esni.md

```diff
@@ -895,13 +895,18 @@ Otherwise, the server reconstructs ClientHelloInner from
 EncodedClientHelloInner, as described in {{encoding-inner}}. It then stops
 iterating over the candidate ECHConfig values.
 
-Upon determining the ClientHelloInner, the client-facing server then forwards
-the ClientHelloInner to the appropriate backend server, which proceeds as in
-{{backend-server}}. If the backend server responds with a HelloRetryRequest,
-the client-facing server forwards it, decrypts the client's second
-ClientHelloOuter using the procedure in {{server-hrr}}, and forwards the
-resulting second ClientHelloInner. The client-facing server forwards all other
-TLS messages between the client and backend server unmodified.
+Upon determining the ClientHelloInner, the client-facing server then checks
+that the message includes the "ech_is_inner" extension, omits the
+"encrypted_client_hello" extension, and does not offer TLS 1.2 or below
+versions. If any of these checks fails, the client-facing server MUST
+abort with an "illegal_parameter" alert. If these checks succeed, the
+client-facing server then forwards the ClientHelloInner to the appropriate
+backend server, which proceeds as in {{backend-server}}. If the backend server
+responds with a HelloRetryRequest, the client-facing server forwards it,
+decrypts the client's second ClientHelloOuter using the procedure in
+{{server-hrr}}, and forwards the resulting second ClientHelloInner. The
+client-facing server forwards all other TLS messages between the client and
+backend server unmodified.
```

Comment:
This is a [wall of text](https://en.uncyclopedia.co/wiki/Wall_of_Text) now.

It might be time to break this down a little more.  Pause for a breath occasionally.

Reply:
👍